src/core/server/http/router/validator.ts

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { SchemaTypeError, ValidationError, ObjectType, Type, TypeOf } from '@kbn/config-schema';
+
+/**
+ * Error to return when the validation is not successful.
+ * @public
+ */
+export class RouteValidationError extends SchemaTypeError {
+  constructor(error: Error | string, path: string[] = []) {
+    super(error, path);
+
+    // Set the prototype explicitly, see:
+    // https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#extending-built-ins-like-error-array-and-map-may-no-longer-work
+    Object.setPrototypeOf(this, RouteValidationError.prototype);
+  }
+}
+
+/**
+ * Allowed returned format of the custom validate function
+ * @public
+ */
+export type RouteValidateFunctionReturn<T> =
+  | {
+      value: T;
+      error?: undefined;
+    }
+  | {
+      value?: undefined;
+      error: RouteValidationError;
+    };
```

Comment:
That was my initial intention, but TS is getting stricter (and in some cases dumber IMHO) with newer versions and with that approach, you can't do 

```ts
result = validationSpec(data);
if (result.error) { // Property 'error' does not exist on type 'RouteValidateFunctionReturn<unknown>'
...
```

There used to be a time when that was possible. 🤦‍♂ 

On top of that, it will increase the complexity of `ReturnType<typeof validationSpec>['value']` because of a similar issue: `'value' might not exist`.

Apparently that's the way to tell TS now it's an object with either one property or the other. There is also another way: `unwantedProp?: never`. I can change it to that if you think it makes more sense.